### PR TITLE
refactor: Add GAZELLE_VERBOSE env and log parser failures

### PR DESCRIPTION
**Pull request by dougthor42**
While investigating #2396 and why #2413 doesn't appear to be working for us, I realized that one of the things I was making heavy use of was additional parser logging that I had added. This adds some of that logging. I also throw in some documentation because I found it helpful.

Users can (attempt to) get additional parse failure information by setting the `GAZELLE_VERBOSE` environment variable to `1`. Eg:

```console
$ GAZELLE_VERBOSE=1 bazel run //:gazelle
```

Here are some example logs:

```console
$ GAZELLE_VERBOSE=1 bazel run //:gazelle
INFO: Invocation ID: a4e026d8-17df-426c-b1cc-d3980690dd53
...
INFO: Running command line: bazel-bin/gazelle
INFO: Streaming build results to: https://btx.cloud.google.com/invocations/a4e026d8-17df-426c-b1cc-d3980690dd53
gazelle: WARNING: failed to parse "hello/get_deps.py". The resulting BUILD target may be incorrect.
gazelle: Parse error at {Row:1 Column:0}:
def search_one_more_level[T]():
gazelle: The above was parsed as: (ERROR (identifier) (call function: (list (identifier)) arguments: (argument_list)))
gazelle: ERROR: failed to generate target "//hello:get_deps" of kind "py_library": a target of kind "pyle_py_binary" with the same name already exists. Use the '# gazelle:python_library_naming_convention' directive to change the naming convention.
$
$ bazel run //:gazelle
INFO: Invocation ID: 726c9fd6-f566-4c30-95ef-c4781ad155de
...
INFO: Running command line: bazel-bin/gazelle
INFO: Streaming build results to: https://btx.cloud.google.com/invocations/726c9fd6-f566-4c30-95ef-c4781ad155de
gazelle: WARNING: failed to parse "hello/get_deps.py". The resulting BUILD target may be incorrect.
gazelle: ERROR: failed to generate target "//hello:get_deps" of kind "py_library": a target of kind "pyle_py_binary" with the same name already exists. Use the '# gazelle:python_library_naming_convention' directive to change the naming convention.
```

### CODE REVIEW

File: gazelle/python/file_parser.go [hunk 1]
```diff
@@ -64,9 +68,27 @@ func ParseCode(code []byte) (*sitter.Node, error) {
 		return nil, err
 	}
 
-	return tree.RootNode(), nil
+	root := tree.RootNode()
+	if root.HasError() {
+		log.Printf("WARNING: failed to parse %q. The resulting BUILD target may be incorrect.", path)
+
+		verbose, envExists := os.LookupEnv("GAZELLE_VERBOSE")
+		if envExists && verbose == "1" {
```
Comment (dougthor42) [inline gazelle/python/file_parser.go:76]: If people are happy with this idea, I will make a more generalized function that can be reused. Maybe even add verbosity levels.

I didn't see any option in the core `gazelle` that allows for different verbosity levels, so if there is one please LMK.

File: gazelle/python/file_parser.go [hunk 2]
```diff
@@ -64,9 +68,27 @@ func ParseCode(code []byte) (*sitter.Node, error) {
...
+			for i := 0; i < int(root.ChildCount()); i++ {
+				child := root.Child(i)
+				if child.IsError() {
+					log.Printf("Parse error at %+v:\n%+v", child.StartPoint(), child.Content(code))
+					log.Printf("The above was parsed as: %v", child.String())
```
Comment (dougthor42) [inline gazelle/python/file_parser.go:81]: Add comments demo'ing the logged message.

File: gazelle/python/file_parser.go [hunk 3]
```diff
@@ -55,7 +56,10 @@ func NewFileParser() *FileParser {
 	return &FileParser{}
 }
 
-func ParseCode(code []byte) (*sitter.Node, error) {
+// ParseCode instantiates a new tree-sitter Parser and parses the python code, returning
+// the tree-sitter RootNode.
+// It print a warning if parsing fails.
```
Comment (rickeylev) [inline gazelle/python/file_parser.go:61]: typo:prints
```suggestion
// It prints a warning if parsing fails.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -61,7 +61,8 @@ Unreleased changes template.
 
 {#v0-0-0-added}
 ### Added
-* Nothing yet.
+* (gazelle): Parser failures will now be logged to the terminal. Additional
+  details can be logged by setting `GAZELLE_VERBOSE=1`.
```
Comment (rickeylev) [inline CHANGELOG.md:65]: You can add this to docs/environment-variables.md, too.

**Review by rickeylev — APPROVED** [review]
I changed the title to refactor, since it isn't providing a feature, more just better diagnostics for debugging failures.

File: gazelle/python/file_parser.go [hunk 5]
```diff
@@ -64,9 +68,27 @@ func ParseCode(code []byte) (*sitter.Node, error) {
...
+		if envExists && verbose == "1" {
+			for i := 0; i < int(root.ChildCount()); i++ {
+				child := root.Child(i)
+				if child.IsError() {
+					log.Printf("Parse error at %+v:\n%+v", child.StartPoint(), child.Content(code))
+					log.Printf("The above was parsed as: %v", child.String())
+				}
+			}
+		}
+	}
```
Comment (aignas) [inline gazelle/python/file_parser.go:85]: nit, it is much more idiomatic to do early returns:
```suggestion
	if !root.HasError() {
		return root, nil
	}
	
	log.Printf("WARNING: failed to parse %q. The resulting BUILD target may be incorrect.", path)

	// TODO: if root.HasError, should we still return err=nil?
	verbose, envExists := os.LookupEnv("GAZELLE_VERBOSE")
	if !envExists || verbose != "1" {
		return root, nil
	}
	
	for i := 0; i < int(root.ChildCount()); i++ {
		child := root.Child(i)
		if child.IsError() {
			log.Printf("Parse error at %+v:\n%+v", child.StartPoint(), child.Content(code))
			log.Printf("The above was parsed as: %v", child.String())
		}
	}
	return root, nil
```

**Review by aignas — APPROVED** [review]
In general LGTM, but have questions if we should return `err=nil` if `root.HasError()`. If the Python file cannot be parsed, why should we continue?

File: gazelle/python/file_parser.go (hunk 5 above)
Comment (dougthor42) [reply gazelle/python/file_parser.go:85]: Thanks! I'll update to that in a followup PR.

**Review by dougthor42 — COMMENTED** [review]
> if we should return err=nil if root.HasError(). If the Python file cannot be parsed, why should we continue?

IMO we _should_ continue on parse errors. It's possible that the parse error is unrelated to anything that gazelle cares about, and in that case the resulting target definition will be correct.

**Pull request merged**